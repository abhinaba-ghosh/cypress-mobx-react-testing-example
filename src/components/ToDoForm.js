import React from 'react';
import { inject, observer } from 'mobx-react';

@inject('store')
@observer
class ToDoForm extends React.Component {
    // const [item, setItem] = useState('');

    constructor(props) {
        super(props);
        this.state = {
            item: '',
        };
    }

    render() {
        return (
            <form
                onSubmit={(e) => {
                    this.props.store.addItem(this.state.item);
                    this.setState({ item: '' });
                    e.preventDefault();
                }}
            >
                <input
                    type="text"
                    placeholder="add your tasks here..."
                    value={this.state.item}
                    onChange={(e) => {
                        this.setState({ item: e.target.value });
                    }}
                ></input>
                <button type="submit" disabled={this.state.item ? false : true}>
                    Add
                </button>
            </form>
        );
    }
}

export default ToDoForm;
