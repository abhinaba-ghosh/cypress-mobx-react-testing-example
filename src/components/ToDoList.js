import React from 'react';
import { observer, inject } from 'mobx-react';

@inject('store')
@observer
class ToDoList extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <ul>
                {this.props.store.items.map((item) => (
                    <li key={item}>
                        {item}{' '}
                        <button
                            onClick={(e) => {
                                this.props.store.deleteItem(item);
                            }}
                        >
                            Mark Complete
                        </button>
                    </li>
                ))}
            </ul>
        );
    }
}

export default ToDoList;
