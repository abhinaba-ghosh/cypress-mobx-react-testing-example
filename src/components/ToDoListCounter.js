import React from 'react';
import { observer, inject } from 'mobx-react';

@inject('store')
@observer
class ToDoListCounter extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        if (this.props.store.itemsCount === 0) {
            return <h1>No Tasks in hand...</h1>;
        }

        return (
            <h1>
                You Have {this.props.store.itemsCount}{' '}
                {this.props.store.itemsCount > 1 ? `Tasks` : `Task`}!
            </h1>
        );
    }
}

export default ToDoListCounter;
