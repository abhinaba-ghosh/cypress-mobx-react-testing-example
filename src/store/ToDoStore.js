import { useLocalObservable } from 'mobx-react';
import { createContext } from 'react';

export const StoreContext = createContext();

const StoreProvider = ({ children }) => {
    const store = useLocalObservable(() => ({
        items: [],
        addItem(item) {
            store.items.push(item);
        },

        deleteItem(item) {
            const index = store.items.indexOf(item);
            store.items.splice(index, 1);
        },
        get itemsCount() {
            return store.items.length;
        },
    }));

    if (window.Cypress) {
        window.store = store;
    }

    return (
        <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
    );
};

export default StoreProvider;
