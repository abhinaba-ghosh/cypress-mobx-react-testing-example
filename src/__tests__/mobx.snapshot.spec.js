/* eslint-disable no-undef */
/* eslint-disable jest/valid-expect-in-promise */
/// <reference types="cypress" />

import { mount } from '@cypress/react';
import React from 'react';
import ToDoForm from '../components/ToDoForm';
import ToDoList from '../components/ToDoList';
import ToDoListCounter from '../components/ToDoListCounter';
import { Provider } from 'mobx-react';
import { ToDoStore } from '../store/ToDoStore';

describe('MobX store snapshot tests', () => {
    beforeEach(() => {
        const store = new ToDoStore();
        mount(
            <Provider store={store}>
                <main>
                    <ToDoForm />
                    <ToDoList />
                    <ToDoListCounter />
                </main>
            </Provider>
        );

        if (window.Cypress) {
            window.store = store;
        }
    });

    it('should snapshot test the mobx store', () => {
        const getStore = () => cy.window().its('store');
        getStore().invoke('addItem', 'Buy milk');
        getStore().invoke('addItem', 'Go shopping');

        // before snapshot test
        getStore()
            .its('getItemDetails')
            .should('have.keys', ['name', 'items']);

        // after snapshot tests
        getStore()
            .its('getItemDetails')
            .then(Object.keys)
            .toMatchSnapshot();
    });
});
