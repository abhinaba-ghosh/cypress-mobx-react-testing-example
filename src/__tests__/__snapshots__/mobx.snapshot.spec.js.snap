exports[`MobX store snapshot tests > should snapshot test the mobx store #0`] =
[
  "name",
  "items"
];
